lib/new_relixir/transaction.ex

```diff
@@ -3,67 +3,37 @@ defmodule NewRelixir.Transaction do
   Records information about an instrumented web transaction.
   """
 
-  defstruct [:name, :start_time]
-
-  @typedoc "A New Relixir transaction context."
-  @opaque t :: %__MODULE__{name: String.t, start_time: :erlang.timestamp}
-
-  @typedoc "The name of a model."
-  @type model :: String.t
-
-  @typedoc "The name of a repository action."
-  @type action :: atom
-
-  @typedoc "The name of a query."
-  @type query :: String.t | {model, action}
-
-  @typedoc "Elapsed time in microseconds."
-  @type interval :: non_neg_integer
+  alias NewRelixir.Collector
 
   @doc """
-  Creates a new web transaction.
-
-  This method should be called just before processing a web transaction.
+  Records the total time of a web transaction.
   """
-  @spec start(String.t) :: t
-  def start(name) when is_binary(name) do
-    %__MODULE__{name: name, start_time: :os.timestamp}
+  @spec record_web(transaction :: binary, elapsed_time :: integer) :: :ok
+  def record_web(transaction, elapsed_time)
+      when is_binary(transaction)
+      and is_integer(elapsed_time) do
+    Collector.record_value({transaction, :total}, elapsed_time)
   end
 
   @doc """
-  Finishes a web transaction.
-
-  This method should be called just after processing a web transaction. It will record the elapsed
-  time of the transaction.
+  Records a database query made in a web transaction.
   """
-  @spec finish(t) :: :ok
-  def finish(%__MODULE__{start_time: start_time} = transaction) do
-    end_time = :os.timestamp
-    elapsed = :timer.now_diff(end_time, start_time)
-
-    record_value!(transaction, :total, elapsed)
+  @spec record_db(transaction :: binary, query :: binary, elapsed_time :: integer) :: :ok
+  def record_db(transaction, query, elapsed_time)
+      when is_binary(transaction)
```

Comment:
any special reason for the guard clauses? it seems that the spec contains this information already and I don't see any other matching function heads. Not asking to delete, just wonder if I'm missing something.

Reply:
The guards here are for type safety, because this can be considered a public API. If this was intended for internal use only we could skip those guards.